### PR TITLE
Hanlde OOM exception and log memory consumptions.

### PR DESCRIPTION
**Pull request by ShanshanHe**
#### Description of change
[//]: # (What are you trying to fix? What did you change)
We've seen OutOfMemory exception being thrown few times from PA logs. This patch is to catch the exception and log the memory consumption.

#### Issue
[//]: # (Having an issue # for the PR is required for tracking purposes. If an existing issue does not exist please create one.)

#### PR reviewer notes
[//]: # (Let us know if there is anything we should focus on.)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by Eruanion — DISMISSED** [review]

File: src/PortingAssistant.Client.Analysis/AnalysisHandler.cs [hunk 1]
```diff
@@ -160,6 +161,12 @@ public class PortingAssistantAnalysisHandler : IPortingAssistantAnalysisHandler
 
                 return solutionAnalysisResult;
             }
+            catch (OutOfMemoryException e)
+            {
+                _logger.LogError("Analyze solution {0} with error {1}", solutionFilename, e);
+                MemoryUtils.LogMemoryConsumption(_logger);
```
Comment (ShanshanHe) [inline src/PortingAssistant.Client.Analysis/AnalysisHandler.cs:167]: @flywanli , @huawenm apart from the regular error logs with exception stacktrace, I'm logging the memory consumption separately with `information` level log, will our ops script capture it and include it in our weekly report?

File: src/PortingAssistant.Client.Common/Utils/MemoryUtils.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿using Microsoft.Extensions.Logging;
+using System;
+using System.Diagnostics;
+
+namespace PortingAssistant.Client.Common.Utils
+{
+    public static class MemoryUtils
+    {
+        public static void LogMemoryConsumption(ILogger logger)
```
Comment (birojnayak) [inline src/PortingAssistant.Client.Common/Utils/MemoryUtils.cs:9]: If we are logging, can we also log below components too a)Size of the Solutions(being assessed). b)Memory usage on various components (used in the application**) c) Log also various generations objects and large objects (if possible)... ** if possible, not sure

**Review by birojnayak — CHANGES_REQUESTED** [review]

File: src/PortingAssistant.Client.Common/Utils/MemoryUtils.cs (hunk 2 above)
Comment (ShanshanHe) [reply src/PortingAssistant.Client.Common/Utils/MemoryUtils.cs:9]: Thanks for your comment @birojnayak. 

For `a)` we can for sure include in this PR, we can get the current solution dir and sum the size of all `.cs` files in it and log it, I'll add this in the current PR. 

For `b)` and `c)`, I'm planning to do some memory profiling and analysis in the next couple of days to determine where the most memory consumption is, once we have the result, I will add logs based on that. wdyt?

**Review by Eruanion — COMMENTED** [review]
I also one comment regarding performance, do you think this might negatively affect performance in a significant way, to include the memory logging on the analysis? It appears that it would be fine but I would want to run some tests on large solutions just to be safe and compte with memory logging and without performance.

**Comment by ShanshanHe** [comment]
> I also one comment regarding performance, do you think this might negatively affect performance in a significant way, to include the memory logging on the analysis? It appears that it would be fine but I would want to run some tests on large solutions just to be safe and compte with memory logging and without performance.

Yeah that's a good point to consider. For `LogMemoryConsumption` it's not holding on to any objects/event handler etc, so once the method call is finished, the relevant objects will be garbage collected, on top of that, the method is only called after OOM exception being thrown, followed by a explicit garbage collection, so I don't see too much of overhead there. But for `LogSolutiontSize`, we're calling the method at the beginning of analysis, it's enumerating all the cs files, if the test solution has lots of cs files under, it might create some overhead, I will do some performance test for `LogSolutiontSize` and see how much memory and time it'll use for a large solution.

**Comment by ShanshanHe** [comment]
> > I also one comment regarding performance, do you think this might negatively affect performance in a significant way, to include the memory logging on the analysis? It appears that it would be fine but I would want to run some tests on large solutions just to be safe and compte with memory logging and without performance.
> 
> Yeah that's a good point to consider. For `LogMemoryConsumption` it's not holding on to any objects/event handler etc, so once the method call is finished, the relevant objects will be garbage collected, on top of that, the method is only called after OOM exception being thrown, followed by a explicit garbage collection, so I don't see too much of overhead there. But for `LogSolutiontSize`, we're calling the method at the beginning of analysis, it's enumerating all the cs files, if the test solution has lots of cs files under, it might create some overhead, I will do some performance test for `LogSolutiontSize` and see how much memory and time it'll use for a large solution.

Hi @Eruanion , I added a unit test against `NopCommerce` solution which has 15119 files and 3308 `.cs` files. Here is some statistics running `LogSolutionSize` 10 times. Quick summary that average time taken to run the method is 200ms for the test solution, and all the short-lived objects can be garbage collected. Since the method is only called one time in each PA run, 200ms seems to be okay given the size of test solution (~140mb after compressed, ~400mb without compression). Too bad github does not allow any large file (>100mb) to be uploaded, I'm not able to commit the unit test with this particular solution. Will change it to a smaller test solution and post the new statistics. Please let me know if the data looks okay to you. Thank you! 

```
   15119 files detected in test solution.
    3308 CS files detected in test solution.
    ----------Iteration 0----------
    202ms to run LogSolutiontSize
    1486kb before LogSolutionSize.
    3414kb after LogSolutionSize.
    1488kb after Garbage Collection
    1928kb allocated during LogSolutionSize.
    1926kb got collected by GC.
    ----------Iteration 1----------
    203ms to run LogSolutiontSize
    1496kb before LogSolutionSize.
    3424kb after LogSolutionSize.
    1490kb after Garbage Collection
    1928kb allocated during LogSolutionSize.
    1934kb got collected by GC.
    ----------Iteration 2----------
    179ms to run LogSolutiontSize
    1498kb before LogSolutionSize.
    3418kb after LogSolutionSize.
    1490kb after Garbage Collection
    1920kb allocated during LogSolutionSize.
    1928kb got collected by GC.
    ----------Iteration 3----------
    180ms to run LogSolutiontSize
    1498kb before LogSolutionSize.
    3418kb after LogSolutionSize.
    1491kb after Garbage Collection
    1920kb allocated during LogSolutionSize.
    1927kb got collected by GC.
    ----------Iteration 4----------
    174ms to run LogSolutiontSize
    1499kb before LogSolutionSize.
    3435kb after LogSolutionSize.
    1491kb after Garbage Collection
    1936kb allocated during LogSolutionSize.
    1944kb got collected by GC.
    ----------Iteration 5----------
    165ms to run LogSolutiontSize
    1499kb before LogSolutionSize.
    3419kb after LogSolutionSize.
    1494kb after Garbage Collection
    1920kb allocated during LogSolutionSize.
    1925kb got collected by GC.
    ----------Iteration 6----------
    221ms to run LogSolutiontSize
    1502kb before LogSolutionSize.
    3422kb after LogSolutionSize.
    1494kb after Garbage Collection
    1920kb allocated during LogSolutionSize.
    1928kb got collected by GC.
    ----------Iteration 7----------
    180ms to run LogSolutiontSize
    1502kb before LogSolutionSize.
    3430kb after LogSolutionSize.
    1494kb after Garbage Collection
    1928kb allocated during LogSolutionSize.
    1936kb got collected by GC.
    ----------Iteration 8----------
    172ms to run LogSolutiontSize
    1502kb before LogSolutionSize.
    3430kb after LogSolutionSize.
    1494kb after Garbage Collection
    1928kb allocated during LogSolutionSize.
    1936kb got collected by GC.
    ----------Iteration 9----------
    174ms to run LogSolutiontSize
    1502kb before LogSolutionSize.
    3430kb after LogSolutionSize.
    1494kb after Garbage Collection
    1928kb allocated during LogSolutionSize.
    1936kb got collected by GC.
    ----------Iteration 10----------
    183ms to run LogSolutiontSize
    1502kb before LogSolutionSize.
    3422kb after LogSolutionSize.
    1494kb after Garbage Collection
    1920kb allocated during LogSolutionSize.
    1928kb got collected by GC.
```

**Review by Eruanion — APPROVED** [review]

**Comment by Eruanion** [comment]
That is awesome and in depth testing. Thanks Shanshan. These results look great the performance impact is negligible.

**Comment by ShanshanHe** [comment]
> That is awesome and in depth testing. Thanks Shanshan. These results look great the performance impact is negligible.

Thank you very much for reviewing the PR. I changed the test solution to a smaller one (total file count is 299, total cs file count is 41), and here is the statistics, average time taken is around `10ms`:

```
    ----------Iteration 0----------
    15ms to run LogSolutiontSize
    1491kb before LogSolutionSize.
    1539kb after LogSolutionSize.
    1468kb after Garbage Collection
    48kb allocated during LogSolutionSize.
    71kb got collected by GC.
    ----------Iteration 1----------
    11ms to run LogSolutiontSize
    1476kb before LogSolutionSize.
    1508kb after LogSolutionSize.
    1469kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    39kb got collected by GC.
    ----------Iteration 2----------
    8ms to run LogSolutiontSize
    1477kb before LogSolutionSize.
    1509kb after LogSolutionSize.
    1470kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    39kb got collected by GC.
    ----------Iteration 3----------
    11ms to run LogSolutiontSize
    1478kb before LogSolutionSize.
    1510kb after LogSolutionSize.
    1471kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    39kb got collected by GC.
    ----------Iteration 4----------
    11ms to run LogSolutiontSize
    1479kb before LogSolutionSize.
    1511kb after LogSolutionSize.
    1471kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    40kb got collected by GC.
    ----------Iteration 5----------
    11ms to run LogSolutiontSize
    1479kb before LogSolutionSize.
    1511kb after LogSolutionSize.
    1473kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    38kb got collected by GC.
    ----------Iteration 6----------
    8ms to run LogSolutiontSize
    1481kb before LogSolutionSize.
    1513kb after LogSolutionSize.
    1473kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    40kb got collected by GC.
    ----------Iteration 7----------
    9ms to run LogSolutiontSize
    1481kb before LogSolutionSize.
    1513kb after LogSolutionSize.
    1473kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    40kb got collected by GC.
    ----------Iteration 8----------
    10ms to run LogSolutiontSize
    1481kb before LogSolutionSize.
    1513kb after LogSolutionSize.
    1473kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    40kb got collected by GC.
    ----------Iteration 9----------
    9ms to run LogSolutiontSize
    1481kb before LogSolutionSize.
    1513kb after LogSolutionSize.
    1473kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    40kb got collected by GC.
    ----------Iteration 10----------
    8ms to run LogSolutiontSize
    1481kb before LogSolutionSize.
    1513kb after LogSolutionSize.
    1477kb after Garbage Collection
    32kb allocated during LogSolutionSize.
    36kb got collected by GC
``` 
                                             MVCMusicStore    |    NopCommerce 
Total # of files              |            299                  |         15119
Total # of cs files         |            41                     |         3308
Average run duration   |            10ms                |         200ms
Memory allocation       |            32kb                |         1928kb

`Total # of files` and `Memory allocation` is almost linear correlated.

File: src/PortingAssistant.Client.Common/Utils/MemoryUtils.cs (hunk 2 above)
Comment (birojnayak) [reply src/PortingAssistant.Client.Common/Utils/MemoryUtils.cs:18]: @ShanshanHe please create an ISSUE and tag here in that case for b and c

Comment (ShanshanHe) [reply src/PortingAssistant.Client.Common/Utils/MemoryUtils.cs:18]: [SIM](https://sim.amazon.com/issues/P49081672)
and corresponding [PR](https://github.com/aws/porting-assistant-dotnet-client/pull/105)

**Pull request merged**